[utils][UpdateLLCTestChecks] Add MIR support to update_llc_test_checks.py.

vpykhtin opened the pull request:
This change enables update_llc_test_checks.py to automatically generate MIR checks for RUN lines that use `-stop-before` or `-stop-after` flags allowing tests to verify intermediate compilation stages (e.g., after instruction selection but before peephole optimizations) alongside the final assembly output.

This resulted from the scenario, when I needed to test two instruction matching patterns where the later pattern in the peepholler reverts the earlier pattern in the instruction selector and distinguish it from the case when the earlier pattern didn't worked at all.

Drawback: this may create very noisy tests.

Note: This is 100% Claude Sonnet 4.5 and I don't feel I understand the change in details, but the change looks so small that I decided to give it a chance.

Example:
```
; NOTE: Assertions have been autogenerated by utils/update_llc_test_checks.py UTC_ARGS: --version 5
; RUN: llc -mtriple=amdgcn -mcpu=gfx90a < %s | FileCheck -check-prefix=GFX90A %s
; RUN: llc -mtriple=amdgcn -mcpu=gfx90a -stop-before=si-conditional-fma-peephole < %s | FileCheck -check-prefix=MIR %s

; Pattern 1: Direct fsub with select
; result = a - (cond ? c : 0.0)
define double @cond_sub_basic(double %a, double %c, i1 %cond) {
; GFX90A-LABEL: cond_sub_basic:
; GFX90A:       ; %bb.0:
; GFX90A-NEXT:    s_waitcnt vmcnt(0) expcnt(0) lgkmcnt(0)
; GFX90A-NEXT:    v_and_b32_e32 v4, 1, v4
; GFX90A-NEXT:    v_cmp_eq_u32_e32 vcc, 1, v4
; GFX90A-NEXT:    v_cndmask_b32_e32 v2, 0, v2, vcc
; GFX90A-NEXT:    v_cndmask_b32_e32 v3, 0, v3, vcc
; GFX90A-NEXT:    v_add_f64 v[0:1], v[0:1], -v[2:3]
; GFX90A-NEXT:    s_setpc_b64 s[30:31]
  ; MIR-LABEL: name: cond_sub_basic
  ; MIR: bb.0 (%ir-block.0):
  ; MIR-NEXT:   liveins: $vgpr0, $vgpr1, $vgpr2, $vgpr3, $vgpr4
  ; MIR-NEXT: {{  $}}
  ; MIR-NEXT:   [[COPY:%[0-9]+]]:vgpr_32 = COPY $vgpr4
  ; MIR-NEXT:   undef [[COPY1:%[0-9]+]].sub1:vreg_64_align2 = COPY $vgpr3
  ; MIR-NEXT:   [[COPY1:%[0-9]+]].sub0:vreg_64_align2 = COPY $vgpr2
  ; MIR-NEXT:   undef [[COPY2:%[0-9]+]].sub1:vreg_64_align2 = COPY $vgpr1
  ; MIR-NEXT:   [[COPY2:%[0-9]+]].sub0:vreg_64_align2 = COPY $vgpr0
  ; MIR-NEXT:   [[V_AND_B32_e32_:%[0-9]+]]:vgpr_32 = V_AND_B32_e32 1, [[COPY]], implicit $exec
  ; MIR-NEXT:   [[V_MOV_B32_e32_:%[0-9]+]]:vgpr_32 = V_MOV_B32_e32 -1074790400, implicit $exec
  ; MIR-NEXT:   [[V_CMP_EQ_U32_e64_:%[0-9]+]]:sreg_64_xexec = V_CMP_EQ_U32_e64 1, [[V_AND_B32_e32_]], implicit $exec
  ; MIR-NEXT:   undef [[V_CNDMASK_B32_e64_:%[0-9]+]].sub1:vreg_64_align2 = V_CNDMASK_B32_e64 0, 0, 0, [[V_MOV_B32_e32_]], [[V_CMP_EQ_U32_e64_]], implicit $exec
  ; MIR-NEXT:   [[V_CNDMASK_B32_e64_:%[0-9]+]].sub0:vreg_64_align2 = AV_MOV_B32_IMM_PSEUDO 0, implicit $exec
  ; MIR-NEXT:   [[COPY2:%[0-9]+]]:vreg_64_align2 = nofpexcept V_FMAC_F64_e32 [[V_CNDMASK_B32_e64_]], [[COPY1]], [[COPY2]], implicit $mode, implicit $exec
  ; MIR-NEXT:   $vgpr0 = COPY [[COPY2]].sub0
  ; MIR-NEXT:   $vgpr1 = COPY [[COPY2]].sub1
  ; MIR-NEXT:   SI_RETURN implicit $vgpr0, implicit $vgpr1
  %sel = select i1 %cond, double %c, double 0.0
  %result = fsub double %a, %sel
  ret double %result
}
```

vpykhtin committed:
[utils][UpdateLLCTestChecks] Add MIR support to update_llc_test_checks.py

This change enables update_llc_test_checks.py to automatically generate
MIR checks for RUN lines that use -stop-before or -stop-after flags.

This allows tests to verify intermediate compilation stages (e.g., after
instruction selection but before peephole optimizations) alongside the
final assembly output.

Remove extra indentation from MIR check lines

conflict processing

arichardson reviewed (approved):
Two minor comments otherwise LGTM. Thanks

arichardson commented:
> Maybe naive question, but for your scenario, can you run just the passes you need, on MIR? I.e you could start with the pattern that you expect to be undone, and not rely on the earlier pass producing it. You'd also not need this change then.
> 
> ... which would avoid teasing the sentiments in this thread:
> 
> https://discourse.llvm.org/t/rfc-llvm-ai-tool-policy-start-small-no-slop/88476
> 
> (I'm assuming @arichardson's lgtm means at least one person is ok with maintaining Claude's output here)

The reason I approved this is that the change looks trivial and I have had cases in the past where I wanted to test the llc output both after a specific pass (usually ISel) as well as the final assembly. Not having to have two tests that could get out of sync is quite nice there but of course you could also use manual check lines or add a MIR test.

vpykhtin commented on llvm/test/tools/UpdateTestChecks/update_llc_test_checks/Inputs/x86_asm_mir_mixed.ll.expected:
```diff
@@ -0,0 +1,47 @@
+; ASM-NEXT:    movl %edi, %eax
+; ASM-NEXT:    addl -{{[0-9]+}}(%rsp), %eax
+; ASM-NEXT:    retq
+  ; MIR-LABEL: name: test2
```
Done.

vpykhtin commented on llvm/utils/update_llc_test_checks.py:
```diff
@@ -16,6 +16,7 @@
 import sys
 
 from UpdateTestChecks import common
+import update_mir_test_checks  # Reuse MIR parsing code.
```
I decided to separate the move into a separate [commit ](https://github.com/llvm/llvm-project/pull/165535) and rebase this PR on top of it.

vpykhtin commented:
Rebased and squashed.

After some testing I made additional changes:
* Modified code to make cleaner separation between `run_list `and `mir_run_list`.
* Added code and test to check conflicts between ASM and MIR prefixes.
* Added test on update of already updated source (I got double checks at some point).

Check indentation looks consistent with ASM checks.

vpykhtin committed:
improved commentary

vpykhtin committed:
Merge branch 'main' into gen_mir_checks_in_ll

arsenm commented on llvm/test/tools/UpdateTestChecks/update_llc_test_checks/Inputs/x86_asm_mir_mixed.ll:
```diff
@@ -0,0 +1,17 @@
+
+define i64 @test1(i64 %i) nounwind readnone {
+  %loc = alloca i64
+  %j = load i64, i64 * %loc
```
Use opaque pointers

arsenm commented on llvm/test/tools/UpdateTestChecks/update_llc_test_checks/Inputs/x86_asm_mir_mixed.ll.expected:
```diff
@@ -0,0 +1,45 @@
+; MIR-NEXT:   $rax = COPY [[ADD64rm]]
+; MIR-NEXT:   RET 0, $rax
+  %loc = alloca i64
+  %j = load i64, i64 * %loc
```
Opaque pointers

vpykhtin committed:
use opaque pointers

vpykhtin committed:
Merge branch 'main' into gen_mir_checks_in_ll

vpykhtin committed:
Merge branch 'main' into gen_mir_checks_in_ll
